### PR TITLE
OCI: add content range handling in patch blob

### LINKED ISSUE #3141: OCI Conformance: "Out-of-order blob upload should return 416"
This was detected running the conformance [test suite](https://github.com/opencontainers/distribution-spec/tree/master/conformance) with Push tests enabled using `OCI_TEST_PUSH=1`.

See [HTML report](https://oci-distribution-conformance-results.s3.amazonaws.com/distribution-split-tests.html)

```
Expected
    <int>: 202
to equal
    <int>: 416
```

```
2020/04/02 15:25:41.959754 DEBUG 
==============================================================================
~~~ REQUEST ~~~
POST  /v2/bloodorange/oci-conformance-test/blobs/uploads/  HTTP/1.1
HOST   : distribution:5000
HEADERS:
	Content-Length: 0
	User-Agent: distribution-spec-conformance-tests
BODY   :
***** NO CONTENT *****
------------------------------------------------------------------------------
~~~ RESPONSE ~~~
STATUS       : 202 Accepted
RECEIVED AT  : 2020-04-02T15:25:41.959370435Z
TIME DURATION: 3.512707ms
HEADERS      :
	Content-Length: 0
	Date: Thu, 02 Apr 2020 15:25:41 GMT
	Docker-Distribution-Api-Version: registry/2.0
	Docker-Upload-Uuid: 04f4260c-b347-4416-a0c0-9aa59c0f360d
	Location: http://distribution:5000/v2/bloodorange/oci-conformance-test/blobs/uploads/04f4260c-b347-4416-a0c0-9aa59c0f360d?_state=*****
	Range: 0-0
	X-Content-Type-Options: nosniff
BODY         :

==============================================================================

2020/04/02 15:25:41.963857 DEBUG 
==============================================================================
~~~ REQUEST ~~~
PATCH  /v2/bloodorange/oci-conformance-test/blobs/uploads/04f4260c-b347-4416-a0c0-9aa59c0f360d?_state=*****  HTTP/1.1
HOST   : distribution:5000
HEADERS:
	Content-Length: 22
	Content-Range: 3-24
	Content-Type: application/octet-stream
	User-Agent: distribution-spec-conformance-tests
BODY   :
bG8sIGhvdyBhcmUgeW91IHRvZGF5Pw==
------------------------------------------------------------------------------
~~~ RESPONSE ~~~
STATUS       : 202 Accepted
RECEIVED AT  : 2020-04-02T15:25:41.963556043Z
TIME DURATION: 3.693907ms
HEADERS      :
	Content-Length: 0
	Date: Thu, 02 Apr 2020 15:25:41 GMT
	Docker-Distribution-Api-Version: registry/2.0
	Docker-Upload-Uuid: 04f4260c-b347-4416-a0c0-9aa59c0f360d
	Location: http://distribution:5000/v2/bloodorange/oci-conformance-test/blobs/uploads/04f4260c-b347-4416-a0c0-9aa59c0f360d?_state=*****
	Range: 0-21
	X-Content-Type-Options: nosniff
BODY         :

==============================================================================

```

### PR DESCRIPTION
**Pull request by wy65701436**
Fixes #3141

1, return 416 for Out-of-order blob upload
2, return 400 for content length and content size mismatch

Signed-off-by: wang yan <wangyan@vmware.com>

### CODE REVIEW

File: registry/api/v2/errors.go
Comment (thaJeztah) [inline registry/api/v2/errors.go:42]: Nit (I see there's other places that should be updated as well); try to avoid using future tense;

```suggestion
		Description: `When a layer is uploaded, the provided range is
		checked against the uploaded chunk. If the range is out of order,
		this error is returned.`,
```

Perhaps a more active voice would be even better ("This error is returned if ...")

File: registry/handlers/api_test.go
Comment (thaJeztah) [inline registry/handlers/api_test.go:2273]: Looks like the `cr` property is used in a couple of places (and not only close to where it's defined here). Perhaps make the property self-describing so that the comment is not needed to understand its purpose.

```suggestion
	contentRange string
```

A comment could still be added to provide more details if you think that's useful, e.g.:

```suggestion
	// Content-Range header to set when pushing chunks
	contentRange string
```

However, also see my other comment below

File: registry/handlers/api_test.go
Comment (thaJeztah) [inline registry/handlers/api_test.go:2276]: I'm guessing you're using `...chunkOption` to make the signature backward compatible. 

However, this function is not exported, and is only able to handle the first option passed (if multiple options are passed, they will be ignored).

If you took this approach to allow future extensions; given that `chunkOption` is a struct, that would already be possible by adding additional properties to `chunkOption`.

I think better options would be to either;

Accept a single option;

```suggestion
func doPushChunk(t *testing.T, uploadURLBase string, body io.Reader, options chunkOptions) (*http.Response, error) {
```

Or, if we don't anticipate further options to be added in the near future, keep it simple, and pass the content-range directly. If further options are added in future, it can be rewritten to use a struct, or functional arguments)

```suggestion
func doPushChunk(t *testing.T, uploadURLBase string, body io.Reader, contentRange string) (*http.Response, error) {
```

File: registry/handlers/api_test.go
Comment (thaJeztah) [inline registry/handlers/api_test.go:2271]: Perhaps `chunkOptions` (plural) is a better name, to indicate this struct could have more than a single option at some point. However, also check my other comment.

**Review by thaJeztah — CHANGES_REQUESTED** [review]
just some quick comments after glancing over the changes

**Comment by wy65701436** [comment]
@thaJeztah thanks for the comments, I have updated the code accordingly.

File: registry/api/v2/errors.go
Comment (wy65701436) [reply registry/api/v2/errors.go]: updated, thanks.

File: registry/handlers/api_test.go
Comment (wy65701436) [reply registry/handlers/api_test.go]: updated, thanks.

File: registry/handlers/api_test.go
Comment (wy65701436) [reply registry/handlers/api_test.go]: I pefer to reserve the options struct as it's easy to support the future extensions.

File: registry/handlers/api_test.go
Comment (wy65701436) [reply registry/handlers/api_test.go]: updated, thanks.

**Comment by xaleeks** [comment]
Is there any update on this?  We are looking forward to this PR.  @thaJeztah

**Review by milosgajdos — APPROVED** [review]
LGTM

**Review by waynr — APPROVED** [review]
:see_no_evil:

**Comment by milosgajdos** [comment]
I think we need a tick from @thaJeztah before we can merge this in because he had requested some changes :)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
changes look good, but can you squash the two commits @wy65701436 ? (I don't think we need the fix-ups as a separate commit in git history)

**Comment by milosgajdos** [comment]
We can squash the commit on merge via `squash and merge` option @thaJeztah. I think we should make that a default setting on this repo, actually.

**Comment by milosgajdos** [comment]
I've changed the merge settings on the repo -- squash and merge is now the default action.

**Comment by thaJeztah** [comment]
> I've changed the merge settings on the repo -- squash and merge is now the default action.

Please don't use that option. It's bad. I would recommend the _reverse_ and disable both the "squash and merge" and "rebase and merge" options.

- commits are an important mechanism to group changes into logical parts. It's important that reviewer and contributor come to the desired grouping of changes during review (which _can be_ "a single commit" for all changes but there are many reasons for grouping otherwise. Commits should _NOT_ be squashed by default)
- using the "squash and commit" option modifies the contributor's commit, and removes GPG signing from the commit. Doing so breaks the chaing of trust (and yes, "spoofing" an author by configuring someone else's e-mail address in your Git config is possible.
- because the "squash and commit" (and "rebase and commit") options modify commits, it's difficult (sometimes "impossible") to correlate commits with their original pull requests; commits in the PR will not show up in git history.

**Comment by thaJeztah** [comment]
>  I've been successfully using it on many other projects I've been involved with but if the preference is in keeping the git commit history and chasing contributors for rebasing that's fine

Yes, I generally find the feature "useful" in repositories where things I mentioned are less important, and where we expect less "tech-savvy" contributors (for example our docs repository, where often contributions are made by editing files through GitHub's UI).
I realise things in this repository have stalled for some time, so there will be a fair share of old PRs that will need rebasing (either by contributor, or a maintainer to carry); hopefully that will only be for a short time (until maintainers have caught up with the backlog).

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
